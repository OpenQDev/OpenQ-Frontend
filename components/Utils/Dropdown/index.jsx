// Third party
import React, { useState, useEffect, useRef } from 'react';
import Exit from '../../svg/exit';
import Check from '../../svg/check';

const Dropdown = ({ toggleFunc, toggleVal, names, title, styles, width, dropdownWidth, removeFunc }) => {
  // Hooks
  const [open, updateOpen] = useState(false);
  const ref = useRef();

  useEffect(() => {
    // Courtesy of https://stackoverflow.com/questions/32553158/detect-click-outside-react-component
    function handleClickOutside(event) {
      if (ref.current && !ref.current.contains(event.target)) {
        updateOpen(false);
      }
    }

    // Bind the event listener
    if (ref.current) {
      document.addEventListener('mousedown', handleClickOutside);
    }
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [ref]);

  // Methods
  const addSelection = (e, name) => {
    if (!toggleVal.includes(name)) toggleFunc(name);
    e.stopPropagation();
  };
  const removeSelection = (e, name) => {
    e.stopPropagation();
    removeFunc(name);
  };

  const handleToggle = (e) => {
    e.preventDefault();
    if (e.target.open) {
      updateOpen(true);
    }
  };

  // Render
  return (
    <details
      ref={ref}
      onToggle={handleToggle}
      open={open}
      className={`h-7 ${styles} text-sm text-muted cursor-pointer`}
    >
      <summary
        className={`inline-flex justify-between ${width} rounded-l-sm border-border-default text-primary rounded-l-smleading-2 py-[5px] px-4`}
      >
        {title || toggleVal}

        <svg
          xmlns='http://www.w3.org/2000/svg'
          className='fill-muted relative top-1'
          viewBox='0 0 16 16'
          width='16'
          height='16'
        >
          <path d='M4.427 7.427l3.396 3.396a.25.25 0 00.354 0l3.396-3.396A.25.25 0 0011.396 7H4.604a.25.25 0 00-.177.427z'></path>
        </svg>
      </summary>
      <div className='w-0'>
        <ul
          className={`z-20 relative bg-subtle ${dropdownWidth} max-h-60 overflow-y-auto mt-2 mb-4 shadow-[rgb(1,_4,_9)_0px_8px_24px_0px] border-web-gray border rounded-sm`}
        >
          {names.length > 0 ? (
            <>
              {names.map((name) => (
                <li className='relative w-full hover:bg-dropdown-hover' key={name}>
                  <button
                    className={` text-left p-2 ${(!removeFunc || !toggleVal.includes(name)) && 'w-full'}`}
                    onClick={(e) => addSelection(e, name)}
                    value={name}
                  >
                    {toggleVal.includes(name) ? (
                      <Check className='inline mr-2  fill-muted' />
                    ) : (
                      <div className='inline-block w-6'></div>
                    )}
                    {name}
                  </button>
                  {toggleVal.includes(name) && removeFunc && (
                    <button onClick={(e) => removeSelection(e, name)}>
                      <Exit className='inline mr-2 fill-muted  absolute right-0 top-3' />
                    </button>
                  )}
                </li>
              ))}{' '}
            </>
          ) : (
            <div className='w-full text-left p-2 px-4'>no label available</div>
          )}
        </ul>
      </div>
    </details>
  );
};
export default Dropdown;
