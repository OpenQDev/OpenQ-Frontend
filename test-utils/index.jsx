// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
