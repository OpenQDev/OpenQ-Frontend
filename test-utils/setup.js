import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
