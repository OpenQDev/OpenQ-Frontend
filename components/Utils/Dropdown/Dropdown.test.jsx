/**
 * @vi-environment jsdom
 */
import React from 'react';
import { render, screen } from '../../../test-utils';
import Dropdown from '.';
import userEvent from '@testing-library/user-event';
import nextRouter from 'next/router';
import renderer from 'react-test-renderer';

nextRouter.useRouter = vi.fn();
describe('Dropdown', () => {
  const mockDropdown = vi.fn();
  const title = 'pizzas';
  beforeEach(() => {});
  const names = ['foo', 'bar'];

  it('should match DOM Snapshot', () => {
    const tree = renderer.create(
      <Dropdown toggleFunc={mockDropdown} title={title} names={names} toggleVal={names[0]} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should be useable dropdown', async () => {
    // ARRANGE
    const user = userEvent.setup();
    render(<Dropdown toggleFunc={mockDropdown} names={names} toggleVal={names[0]} />);
    const defaultTitle = screen.getAllByText(names[0])[0];
    const altName = screen.getByText(names[1]);
    await user.click(defaultTitle);
    await user.click(altName);
    // ACT
    expect(mockDropdown).toHaveBeenCalledWith(names[1]);
  });
});
