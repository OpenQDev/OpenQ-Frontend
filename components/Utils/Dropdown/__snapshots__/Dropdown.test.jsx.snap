// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown > should match DOM Snapshot 1`] = `
<details
  className="h-7 undefined text-sm text-muted cursor-pointer"
  onToggle={[Function]}
  open={false}
>
  <summary
    className="inline-flex justify-between undefined rounded-l-sm border-border-default text-primary rounded-l-smleading-2 py-[5px] px-4"
  >
    pizzas
    <svg
      className="fill-muted relative top-1"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4.427 7.427l3.396 3.396a.25.25 0 00.354 0l3.396-3.396A.25.25 0 0011.396 7H4.604a.25.25 0 00-.177.427z"
      />
    </svg>
  </summary>
  <div
    className="w-0"
  >
    <ul
      className="z-20 relative bg-subtle undefined max-h-60 overflow-y-auto mt-2 mb-4 shadow-[rgb(1,_4,_9)_0px_8px_24px_0px] border-web-gray border rounded-sm"
    >
      <li
        className="relative w-full hover:bg-dropdown-hover"
      >
        <button
          className=" text-left p-2 w-full"
          onClick={[Function]}
          value="foo"
        >
          <svg
            className="inline mr-2  fill-muted"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z"
              fillRule="evenodd"
            />
          </svg>
          foo
        </button>
      </li>
      <li
        className="relative w-full hover:bg-dropdown-hover"
      >
        <button
          className=" text-left p-2 w-full"
          onClick={[Function]}
          value="bar"
        >
          <div
            className="inline-block w-6"
          />
          bar
        </button>
      </li>
       
    </ul>
  </div>
</details>
`;
